### ISSUE TITLE
agnostic unable to find mysqldump despite being present in PATH

### ISSUE THREAD

**Issue by karthicraghupathi** [TeamHG-Memex/agnostic#14]
It appears that `agnostic` is unable to find `mysqldump` despite it being present in PATH. This is happening to me on macOS Catalina using Python 3.6.7.

I created a new `virtualenv` using `pyenv` and installed `agnostic`. I then attempted to create a snapshot of an existing DB:

```
00:38 karthicr@ws1:agnostic_test $ pyenv virtualenv 3.6.7 agnostic_test
Looking in links: /var/folders/yq/82xt_rsn3ds1473grw012dqxtx0bq9/T/tmp0qi57336
Requirement already satisfied: setuptools in /Users/karthicr/.pyenv/versions/3.6.7/envs/agnostic_test/lib/python3.6/site-packages (39.0.1)
Requirement already satisfied: pip in /Users/karthicr/.pyenv/versions/3.6.7/envs/agnostic_test/lib/python3.6/site-packages (10.0.1)

00:39 karthicr@ws1:agnostic_test $ pyenv local agnostic_test

(agnostic_test) 00:39 karthicr@ws1:agnostic_test $ pip install agnostic[mysql]
Looking in indexes: https://pypi.python.org/simple
Collecting agnostic[mysql]
  Using cached https://files.pythonhosted.org/packages/35/7f/30dfb7689a58ff1a2f9bf9ac9750c21e08744b96ad1490fd2d3407fbf480/agnostic-1.0.1.tar.gz
Collecting Click<8.0,>=7.0 (from agnostic[mysql])
  Using cached https://files.pythonhosted.org/packages/fa/37/45185cb5abbc30d7257104c434fe0b07e5a195a6847506c074527aa599ec/Click-7.0-py2.py3-none-any.whl
Collecting sqlparse<0.3.0,>=0.2.4 (from agnostic[mysql])
  Using cached https://files.pythonhosted.org/packages/65/85/20bdd72f4537cf2c4d5d005368d502b2f464ede22982e724a82c86268eda/sqlparse-0.2.4-py2.py3-none-any.whl
Collecting PyMySQL<0.10.0,>=0.9.2 (from agnostic[mysql])
  Using cached https://files.pythonhosted.org/packages/ed/39/15045ae46f2a123019aa968dfcba0396c161c20f855f11dea6796bcaae95/PyMySQL-0.9.3-py2.py3-none-any.whl
Installing collected packages: Click, sqlparse, PyMySQL, agnostic
  Running setup.py install for agnostic ... done
Successfully installed Click-7.0 PyMySQL-0.9.3 agnostic-1.0.1 sqlparse-0.2.4

(agnostic_test) 00:39 karthicr@ws1:agnostic_test $ pip freeze
agnostic==1.0.1
Click==7.0
PyMySQL==0.9.3
sqlparse==0.2.4

(agnostic_test) 00:40 karthicr@ws1:agnostic_test $ mkdir migrations

(agnostic_test) 00:40 karthicr@ws1:agnostic_test $ agnostic -t mysql -u karthicr -d playground snapshot current.sql
Enter password for "karthicr" on "playground":
Creating snapshot of database [playground]…
Error: Not able to create snapshot: [Errno 2] No such file or directory: 'mysqldump': 'mysqldump'

(agnostic_test) 00:40 karthicr@ws1:agnostic_test $ python --version
Python 3.6.7

(agnostic_test) 00:43 karthicr@ws1:agnostic_test $ pip --version
pip 19.3.1 from /Users/karthicr/.pyenv/versions/3.6.7/envs/agnostic_test/lib/python3.6/site-packages/pip (python 3.6)
```

As you can see from the above, I get: `Error: Not able to create snapshot: [Errno 2] No such file or directory: 'mysqldump': 'mysqldump'`.

However I can confirm `mysqldump` is available and present in the PATH and running the snapshot command manually results in an output.

```
(agnostic_test) 00:43 karthicr@ws1:agnostic_test $ which mysql
/usr/local/opt/mysql@5.7/bin/mysql

(agnostic_test) 00:43 karthicr@ws1:agnostic_test $ which mysqldump
/usr/local/opt/mysql@5.7/bin/mysqldump

(agnostic_test) 00:44 karthicr@ws1:agnostic_test $ mysqldump -h localhost -u karthicr -p --no-create-db --no-data --compact playground
Enter password:
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!40101 SET character_set_client = utf8 */;
CREATE TABLE `agnostic_migrations` (
  `name` varchar(255) COLLATE utf8mb4_unicode_ci NOT NULL,
  `status` varchar(255) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `started_at` timestamp NULL DEFAULT NULL,
  `completed_at` timestamp NULL DEFAULT NULL,
  PRIMARY KEY (`name`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
/*!40101 SET character_set_client = @saved_cs_client */;
```

I was going through the code and I feel like the issue is occurring because of the `env` dictionary at https://github.com/TeamHG-Memex/agnostic/blob/5c731a3f60e00f286f8ff06303baa31592b0d8cb/agnostic/mysql.py#L110. 

I understand that this is necessary to pass the password to the underlying `mysqldump` call. However per Python 3 documentation at https://docs.python.org/3/library/subprocess.html:

> If env is not None, it must be a mapping that defines the environment variables for the new process; these are used instead of the default behavior of inheriting the current process’ environment. It is passed directly to Popen.

So when the `env` dictionary is passed as a parameter to `subprocess.Popen`, it is overwriting the existing environment variables and as a result `agnostic` fails.

At least that is what I think is happening. Looking forward to your thoughts.

**Comment by karthicraghupathi** [comment]
I may have a possible solution. I'm appending the current environment variables to the `env` dictionary before passing it as parameter to `subprocess.run`.  Here is a sample script:

```
import os
import subprocess

host = 'localhost'
username = 'xx'
password = 'xx'
database = 'testbed'

env = { 'MYSQL_PWD': password }
# add the current environment variables
env.update(os.environ)

command = [
    'mysqldump',
    '-h', host,
    '-u', username,
    '--no-create-db',
    '--no-data',
    '--compact',
    database
]

result = subprocess.run(command, env=env, stderr=subprocess.PIPE)
```

If you think this is a good way to solve the problem, I can send you a pull request with this change.

Let me know.

**Comment by karthicraghupathi** [comment]
I have submitted a PR for this issue: https://github.com/TeamHG-Memex/agnostic/pull/15

**Comment by mehaase** [comment]
Closed via #15
